File: scripts/gather-new-data/02_export_ecclimate_csv.R
# Author: Saeesh Mangwani
# Date: 2022-06-15

# Description: Exporting EC Climate data to CSV format based on a provided
# date range

# ==== Libraries ====
library(DBI)
library(RPostgres)
library(rjson)
library(readr)
library(optparse)
library(lubridate)
source('scripts/help_funcs.R')

# ==== Initializing option parsing ====
option_list <-  list(
  make_option(c("-s", "--startdate"), type="character", default=(Sys.Date() - 30), 
              help="A year month combination indicating the start date for data download. [Default= %default]", 
              metavar="character")
)

# Parse any provided options and store them in a list
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

# ==== Paths and global variables ====

# Path to credentials
dbase_credentials_path <- 'credentials.json'

# Path to exported csvs
csv_path <- 'data/csv/ecclimate'
dir_check_create(csv_path)

# ==== Opening database connection ====

# Credentials files
creds <- fromJSON(file = 'credentials.json')

# Opening database connection
conn <- dbConnect(RPostgres::Postgres(),
                  host = creds$host, dbname = creds$dbname,
                  user = creds$user, password = creds$password)

# ==== Daily data ====

# Data query
query <- format_simple_query('ecclimate', 'daily', 
                             date_col = 'datetime', start_date = ymd(opt$startdate))

# Reading data
daily <- dbGetQuery(conn, query)

# Writing to csv
write_csv(daily, file.path(csv_path, 'ecclimate-daily.csv'))
rm(daily)
gc()

# ==== Hourly data ====

# Data query
# query <- format_simple_query('ecclimate', 'hourly', 
#                              date_col = 'datetime', start_date = ymd(opt$startdate))
# 
# # Reading data
# hourly <- dbGetQuery(conn, query)
# 
# # Writing to csv
# write_csv(hourly, file.path(csv_path, 'ecclimate-hourly.csv'))

# Closing database connection
dbDisconnect(conn)

File: scripts/gather-new-data/02_export_ecclimate_csv.py
# Author: Saeesh Mangwani
# Date: 02/08/2023

# Description: Exporting EC Climate data to CSV format based on a provided date range

# %% ===== Loading libraries =====
import os
import sys
from pathlib import Path
os.chdir(Path(__file__).parent.parent.parent)
sys.path.append(os.getcwd())
from json import load
from optparse import OptionParser
from datetime import datetime, timedelta
import psycopg2
import pandas as pd

#%% Initializing option parsing
parser = OptionParser()
parser.add_option(
    "-s", "--startdate", 
    dest="startdate",
    default=(datetime.today() - timedelta(days=31)).strftime("%Y-%m-%dT00:00:00-00:00"),
    help="The start date of the date range for which data are being posted. Defaults to 31 days before today")
parser.add_option(
    "-e", "--enddate", 
    dest="enddate",
    default=datetime.today().strftime("%Y-%m-%dT00:00:00-00:00"),
    help="The end date of the date range for which data are being posted. Defaults to today")
(options, args) = parser.parse_args()

# %% ===== Paths and global variables =====

# Client credentials from JSON
creds = load(open('options/dbase_credentials.json',))

# Filepaths
fpaths = load(open('options/filepaths.json', ))

# Ensuring directory exists for holding posting data data
out_dir = fpaths['update-data-dir']
if not os.path.exists(out_dir):
    os.makedirs(out_dir)

# %% ===== Initializing database connection =====

# Database connection
conn = psycopg2.connect(
    host=creds['host'], 
    port=creds['port'],
    database=creds['dbname'],
    user= creds['user'],
    password=creds['password']
)
cursor = conn.cursor()

# %% ===== Initializing update options =====
start_date = options.startdate
end_date = options.enddate
# start_date = (datetime.today() - timedelta(days=31)).strftime("%Y-%m-%dT00:00:00-00:00")
# end_date =datetime.today().strftime("%Y-%m-%dT00:00:00-00:00")

# %% ==== Gathering update data ====

# Query options
schema = 'ecclimate'
table = 'daily'
datecol = 'datetime'
# Building query
query = 'select * from {0}.{1} where "{2}" >= \'{3}\' and "{2}" <= \'{4}\''.format(
    schema,
    table,
    datecol,
    start_date,
    end_date
)

# Getting data
cursor.execute(query)
col_names = [i[0] for i in cursor.description]
daily = pd.DataFrame(cursor.fetchall(), columns=col_names)


# %%  ==== Exporting to CSV ====
if daily.shape[0] == 0:
    print("No new data available for EC-Climate between {} and {}. No CSV exported".format(start_date, end_date))
else:
    print("Exporting data to CSV")
    daily.to_csv(out_dir + '/ecclimate-daily.csv', index=False)



